_posts/2020-09-11-RestTemplate-vs-WebClient.md

```diff
@@ -0,0 +1,556 @@
+
+### Reactive Database Drivers
+
+There have been supporting libraries for Reactive NoSQL for a while, which can retrieve data from the database in a reactive manner. 
+Which means they start returning data as the database is querying, and not when the entire database has been queried. So you are good to go. 
+But, with Relational databases (which you are probably using as well), there weren’t any in Spring until recently. 
+Before, you could wrap your database response in `Flux` and `Mono` wrappers as soon as possible when calling your database, but if these databases are the major sources of your data and not just a small portion, it kind of ruins the appeal and benefit. 
+
+### R2DBC: 2 steps forward 1 step back
+
+Not until very recently (December 2019), `spring-data-r2dbc` dependency got released. 
+R2DBC stands for **Reactive Relational DataBase Connectivity**. Bear in mind that this driver is not an ORM. 
+For now, it is not yet capable of mapping relations (eg: `@OneToMany`, `@ManyToOne`, etc). 
+You would have to manage this yourself. However, with this driver, we can query data as streams.
+
+## End-to-end Reactive example
+
+We've learned about WebClient, Reactive streams and R2DBC. Let's use these in an example to really see how it works. 
+We are going to create a 2 microservices with a R2DBC connecting to MySQL databases and a Rest API. 
+Then we'll create another microservices which consumes both of these Rest API's. 
+Finally, an Angular frontend that consumes that microservice to display the data.
+
+### Recipe Service (Reactive R2DBC)
+
+Let’s dive into our Recipe Rest service and see how we can create a reactive rest service. 
+We add the following dependencies.
+
+```xml
+<dependencies>
+    <dependency>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-webflux</artifactId>
+    </dependency>
+    
+    <dependency>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-data-r2dbc</artifactId>
+    </dependency>
+    <dependency>
+        <groupId>dev.miku</groupId>
+        <artifactId>r2dbc-mysql</artifactId>
+        <scope>runtime</scope>
+    </dependency>
+    <dependency>
+        <groupId>mysql</groupId>
+        <artifactId>mysql-connector-java</artifactId>
+        <scope>runtime</scope>
+    </dependency>
+</dependencies>
+```
+
+We configure the database connection and enable R2DBC. 
+
+```java
+@Configuration
+@EnableR2dbcRepositories
+public class ReactiveDatabaseConfig extends AbstractR2dbcConfiguration {
+    @Bean
+    @Override
+    public ConnectionFactory connectionFactory() {
+        return ConnectionFactories.get(ConnectionFactoryOptions
+                .builder()
+                .option(DRIVER, "mysql")
```

Comment:
No, R2DBC provides the ConnectionFactory, but the database is MySQL so you still need the driver for a MySQL database in this case. Maybe I've written something which could be confusing, so I'll re-read that part of the text and edit it if necessary.

Reply:
A MySql driver is indeed still needed. But there is a r2dbc MySql driver that you can use instead of the jdbc MySql driver. I kind of doubt that using r2dbc connections with a jdbc driver is possible, but even if it is it would be a bit useless. You would use the reactive programming model, but the driver would still be blocking and not reactive, which defeats the purpose a bit. 
Is it correct what I'm saying or did I miss something in my research on this?

Reply:
Ah I see the confusion, the name 'DRIVER' doesn't really stand for driver. It doesn't use a JDBC driver, but R2DBC indeed, since it resolves to the connection string: r2dbc:mysql://127.0.0.1/recipe... and not jdbc:mysql...

That's why, it does require 'mysql' to be filled in for 'DRIVER', otherwise this error appears: 'Unable to create a ConnectionFactory for 'ConnectionFactoryOptions{options={driver=r2dbc, password=REDACTED, database=recipe, host=127.0.0.1, user=root}}'. Available drivers: [ pool, mysql ]'